client/tso_mcs_discovery.go

```diff
@@ -0,0 +1,306 @@
+
+	// Client option.
+	option *option
+}
+
+// newTSOMcsDiscovery returns a new BaseClient of a TSO microservice.
+func newTSOMcsDiscovery(ctx context.Context, cancel context.CancelFunc, wg *sync.WaitGroup, metacli MetaStorageClient,
+	keyspaceID uint32, urls []string, tlsCfg *tlsutil.TLSConfig, option *option) ServiceDiscovery {
+	bc := &tsoMcsDiscovery{
+		ctx:               ctx,
+		cancel:            cancel,
+		wg:                wg,
+		metacli:           metacli,
+		keyspaceID:        keyspaceID,
+		primaryKey:        path.Join(tsoKeyspaceGroupPrimaryElectionPrefix+fmt.Sprintf("%05d", 0), "primary"),
+		tlsCfg:            tlsCfg,
+		option:            option,
+		checkMembershipCh: make(chan struct{}, 1),
+	}
+	bc.urls.Store(urls)
+
+	return bc
+}
+
+// Init initialize the concrete client underlying
+func (c *tsoMcsDiscovery) Init() error {
+	if err := c.initRetry(c.updateMember); err != nil {
+		c.cancel()
+		return err
+	}
+	c.wg.Add(1)
+	go c.startCheckMemberLoop()
+	return nil
+}
+
+func (c *tsoMcsDiscovery) initRetry(f func() error) error {
+	var err error
+	for i := 0; i < c.option.maxRetryTimes; i++ {
+		if err = f(); err == nil {
+			return nil
+		}
+		select {
+		case <-c.ctx.Done():
+			return err
+		case <-time.After(time.Second):
+		}
+	}
+	return errors.WithStack(err)
+}
+
+func (c *tsoMcsDiscovery) startCheckMemberLoop() {
+	defer c.wg.Done()
+
+	ctx, cancel := context.WithCancel(c.ctx)
+	defer cancel()
+
+	for {
+		select {
+		case <-c.checkMembershipCh:
+		case <-time.After(memberUpdateInterval):
+		case <-ctx.Done():
+			return
+		}
+		if err := c.updateMember(); err != nil {
+			log.Error("[pd(tso)] failed to update member", errs.ZapError(err))
+		}
+	}
+}
+
+// Close releases all resources
+func (c *tsoMcsDiscovery) Close() {
+	c.clientConns.Range(func(key, cc interface{}) bool {
+		if err := cc.(*grpc.ClientConn).Close(); err != nil {
+			log.Error("[pd(tso)] failed to close gRPC clientConn", errs.ZapError(errs.ErrCloseGRPCConn, err))
+		}
+		c.clientConns.Delete(key)
+		return true
+	})
+}
+
+// GetClusterID returns the ID of the cluster
+func (c *tsoMcsDiscovery) GetClusterID(context.Context) uint64 {
+	return 0
+}
+
+// GetURLs returns the URLs of the servers.
+// For testing use. It should only be called when the client is closed.
+func (c *tsoMcsDiscovery) GetURLs() []string {
+	return c.urls.Load().([]string)
+}
+
+// GetServingAddr returns the grpc client connection of the serving endpoint
+// which is the primary in a primary/secondy configured cluster.
+func (c *tsoMcsDiscovery) GetServingEndpointClientConn() *grpc.ClientConn {
+	if cc, ok := c.clientConns.Load(c.getPrimaryAddr()); ok {
+		return cc.(*grpc.ClientConn)
+	}
+	return nil
+}
+
+// GetClientConns returns the mapping {addr -> a gRPC connectio}
+func (c *tsoMcsDiscovery) GetClientConns() *sync.Map {
+	return &c.clientConns
+}
+
+// GetServingAddr returns the serving endpoint which is the primary in a
+// primary/secondy configured cluster.
```

Comment:
```suggestion
// primary/secondary configured cluster.
```

Reply:
Seems that a lot of places have time problem because of copy-paste. Changed them all.
This is very deliberate code review! Thank you very much!